src/garbage_collector.rs

```diff
@@ -55,25 +62,48 @@ impl GarbageCollector {
                 external_rx,
                 db_stats,
             };
-            tokio_handle.block_on(orchestrator.run());
+            tokio_handle.block_on(orchestrator.run(shutdown_tx));
         });
         Self {
             main_thread: Some(main_thread),
             main_tx: external_tx,
+            shutdown_rx,
+        }
+    }
+
+    /// Waits for the Garbage Collector to complete its GC loops
+    pub(crate) async fn shutdown_notified(mut self) {
```

Comment:
This is how I originally tried to do it, but it turned out to be quite complicated to get right with Rust's synchronization primitives... this was way simpler. It's effectively the same as existing code because `close` will notify the main loop to exit, at which point `shutdown_notified` will be unblocked as well.

The reason it was complicated was because of joining the thread takes ownership over it, which means it has to remove it from the collector instance (by calling `Option##take` or something like that). Handling that in a `Mutex` and `Arc` seemed like overkill for this, which was much easier to get right.

LMK if you feel strongly about this, but I think it's effectively the same behavior as what you suggest. I can even rename `close` `notify_shutdown` and `notified_shutdown` -> `await_shutdown` (I named it `notified_shtudown` to stay in line with Rust's `condvar.notified().await` naming scheme)

Reply:
The other difference between calling `join` and what I have here is that using channels for this makes it `async` whereas `join` is blocking - so it just felt slightly more idiomatic to call `shutdown_notified().await` instead of introducing a blocking call.

Reply:
hmm, I guess if we also want a signal handler to call notify_shutdown we'll have to go through the mutex path anyway since there would be two threads potentially calling `close`/`shutdown`... this rust async stuff is a pain in conjunction with the ownership model, and I haven't actually seen much builtin safety wrt deadlocks.

Reply:
👍 I'll give this a shot and see what I can do with it - thanks for bearing with me why I figure out all this rust async stuff

Reply:
I haven't read over this notifier code in detail (it got pretty complex). I wonder, can we use a watch for `is_shutdown` similar to what I did for `is_flushed` in `mem_table`?

https://docs.rs/tokio/latest/tokio/sync/watch/fn.channel.html

```rust
    pub(crate) async fn await_flush_to_l0(&self) {
        let mut rx = self.is_flushed_rx.clone();
        while !*rx.borrow_and_update() {
            rx.changed().await.expect("watch channel closed");
        }
    }

    pub(crate) fn notify_flush_to_l0(&self) {
        self.is_flushed_tx.send(true).expect("watch channel closed");
    }
```

Reply:
OK, looking at your code the thing you did differently from my prototype is to ignore errors on `main_tx.send(Shutdown)` in `notifyShutdown`. I had wrapped that in `!main_thread.is_finished()` because otherwise we'd try to send a message on a closed stream, and I wanted any other error to throw. If we're OK with that, this makes it much easier because `notify_shutdow`n doesn't need to take ownership of the main thread lock to check that (that's where I was running into the complexity and potential race conditions/deadlocks)